src/actions/product-db-actions.ts

```diff
@@ -0,0 +1,91 @@
+import {Product} from '../db-schemas/product-schema';
+import {ProductDocument} from '../db-schemas/product-document';
+import {logger} from '../logger/logger';
+
+export const retrieveAllProducts = async () => {
+    const products = await Product.find();
+    return products.map((product) => product.toClient());
+};
+
+export const retrieveSpecificProduct = async (id: string) => {
+    const product = await Product.findById(id);
+    if (!product)
+        return null;
+    return product.toClient();
+};
+
+export const createProduct = async (productJson: object) => {
+    const productToCreate = new Product(productJson);
+    const newProduct = await productToCreate.save();
+    return newProduct.toClient();
+};
+
+export const updateProduct = async (id: string, newProductJson: object) => {
+    const modifiedProduct = await Product.findByIdAndUpdate(id, newProductJson, {new: true});
+    if (!modifiedProduct)
+        return null;
+    return modifiedProduct.toClient();
+};
+
+export const deleteProduct = async (id: string) => {
+    const deletedProduct = await Product.findByIdAndDelete(id);
+    if (!deletedProduct)
+        return null;
+    return deletedProduct.toClient();
+};
+
+const processOneProduct = async (productToBuy: { id: string, quantity: number }, existingProduct: ProductDocument, resList: { id: string, success: boolean, error?: string }[]) => {
```

Comment:
in JS we don't have out parameters and don't change any parameter that we get! if you need to change some of the parameters - create copy.
there just return the result, don't change the parameters

Reply:
removed it